### PR TITLE
bump-formula-pr: more enhancements.

### PR DESCRIPTION
**Pull request by ilovezfs**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

- improve formula name guessing
- update GNU mirror handling
- add Debian mirror handling
- detect download strategy from url
- regex escape urls in replacements
- improve version formula component regex

### CODE REVIEW

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 1]
```diff
@@ -138,7 +138,12 @@ def bump_formula_pr
     new_url = ARGV.value("url")
     if new_url && !formula
       is_devel = ARGV.include?("--devel")
-      base_url = new_url.split("/")[0..4].join("/")
+      new_url_split = new_url.split("/")
+      base_url = if new_url_split.count > 5
+        new_url_split[0..4].join("/")
+      else
+        new_url_split[0..(new_url_split.count - 2)].join("/")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: With all these splits a comment would be cool explaining what's going on here (as I have no idea). Perhaps using some/a new variables for the `5`, `4`, `new_url_split.count - 2`?

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -184,14 +189,13 @@ def bump_formula_pr
     elsif !new_url
       odie "#{formula}: no --url= argument specified!"
     else
-      resource_url = if requested_spec != :devel && new_url =~ /.*ftpmirror.gnu.*/
-        new_mirror = new_url.sub "ftpmirror.gnu.org", "ftp.gnu.org/gnu"
-        new_mirror
-      else
-        new_url
+      new_mirror = if requested_spec != :devel && new_url =~ %r{.*ftp.gnu.org/gnu.*}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Could maybe make this a `case` statement if you anticipate more of these in future.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 3]
```diff
@@ -260,7 +264,7 @@ def bump_formula_pr
       end
     elsif forced_version && forced_version == "0"
       if requested_spec == :stable
-        replacement_pairs << [/^  version \"[a-z\d+\.]+\"\n/m, ""]
+        replacement_pairs << [/^  version \"[a-zA-Z\d+\.\-\_]+\"\n/m, ""]
```
Comment (alyssais) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Could be shortened to this, I think:

```ruby
/^  version \"[\w\.\-]+\"\n/m
```

**Comment by ilovezfs** [comment]
PR refreshed.

**Pull request merged**